File: Leetcode/Three/50. Pow(x, n).java
/*
    https://leetcode.com/discuss/52800/5-different-choices-when-talk-with-interviewers
*/
public class Solution {
    public double myPow(double x, int n) {
        boolean isPositive = n >= 0;
        if(!isPositive) n = Math.abs(n);
        double r = help(x, n);
        return isPositive?r:1.0/r;
    }
    private double help(double x, int n) {
        if(n == 0) return 1.0;
        double r = myPow(x, n >> 1);
        if((n & 1) == 1) {
            return r*r*x;
        }
        else return r*r;
    }
}

File: Leetcode/Javascript/50. Pow(x, n).js
var myPow = function(x, n) {
    if(n < 0) return 1/help(x, Math.abs(n));
    else return help(x, n);
};

var help = function(x, n) {
    if(!n) return 1;
    if(n === 1) return x;
    var val = myPow(x, Math.floor(n/2));
    if((n&1)) return val*val*x;
    else return val*val;
}
